### PR TITLE
CRI: Add field to support CPU affinity on Windows

### PR DESCRIPTION
**Pull request by kiashok**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR adds CRI field 'affinity_cpus' to WindowsContainerResources struct to support CPU affinity on windows.
This field will be used by Windows CPU manager to set the logical processors to affinitize for a particular container down to containerd/hcsshim.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?
No

```release-note
CRI: Add field to support CPU affinity on Windows
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig node
/sig windows

/cc @jsturtevant @marosset @aravindhp

### CODE REVIEW

**Comment by aravindhp** [comment]
@kiashok thank you for the PR.

>This field will be used by Windows CPU manager to set the logical processors to affinitize for a particular container down to containerd/hcsshim.

Is any follow up work required for this? It will help if you can link to any docs about the Windows CPU manager.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
Comment (jsturtevant) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1023]: I believe we are going to need a structure like https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-group_affinity, so we can support machines with groups > 64 processors

**Comment by jsturtevant** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/125262

**Comment by ffromani** [comment]
just curious why close? is there a better direction for this work?

**Comment by kiashok** [comment]
> just curious why close? is there a better direction for this work?

@ffromani all I wanted to do is push new changes and when I did so, it closed the PR automatically. Not sure why! Would ideally want to reopen this PR

**Comment by ffromani** [comment]
> > just curious why close? is there a better direction for this work?
> 
> @ffromani all I wanted to do is push new changes and when I did so, it closed the PR automatically. Not sure why! Would ideally want to reopen this PR

If you delete the remote branch (`git push -d origin BRANCHNAME`) it will automatically close the PR. Not sure this is what happened for you but I managed to shot myself in the foot many times with this.

**Comment by kiashok** [comment]
> > > just curious why close? is there a better direction for this work?
> > 
> > @ffromani all I wanted to do is push new changes and when I did so, it closed the PR automatically. Not sure why! Would ideally want to reopen this PR
> 
> If you delete the remote branch (`git push -d origin BRANCHNAME`) it will automatically close the PR. Not sure this is what happened for you but I managed to shot myself in the foot many times with this.

No, I did not delete the remote branch at all. Did a rebase, made new changes and pushed the branch.

**Comment by kiashok** [comment]
> /reopen

Thanks for reopening this PR!

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
Comment (haircommander) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1052]: I think this could be WindowsCpuAffinity and have the same meaning, but I'll defer to your expertise.

Comment (kiashok) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1052]: Named it so to keep it in line with this struct https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/miniport/ns-miniport-_group_affinity . I'll add in a comment to point to this struct in the proto file - is that ok?

Comment (haircommander) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1052]: wfm :)

**Comment by jsturtevant** [comment]
> also: do we need a KEP to add support for windows in the cpu manager?

I wasn't sure so gave it a try to see how much would change: https://github.com/kubernetes/kubernetes/pull/125296.  There really isn't any changes to the any of the related managers and more just wiring it in properly.  Thoughts?

**Comment by haircommander** [comment]
I kinda lean towards one but what do @SergeyKanzhelev and @mrunalp think? It is less code than the[ cgroup driver from cri kep](https://github.com/kubernetes/enhancements/issues/4033) which required one

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1053]: nit: you don't need `en-us` in this link. It will be more user friendly without the language

**Comment by haircommander** [comment]
hey @kiashok @jsturtevant we chatted about this in sig node, and we do want there to be  KEP, but it doesn't need to be a new one. Are you willing to make an update to the [now GA](https://github.com/kubernetes/enhancements/issues/3570) cpu manager KEP and add the approach for windows cpu manager? then we can come to consensus on how it all works, have somewhere to reference, and then we can move forward here. It doesn't need to follow normal KEP deadlines or be a new stage, just having somewhere to talk about approach and differences between windows and linux will be helpful. WDYT

**Comment by jsturtevant** [comment]
> hey @kiashok @jsturtevant we chatted about this in sig node, and we do want there to be KEP, but it doesn't need to be a new one. Are you willing to make an update to the [now GA](https://github.com/kubernetes/enhancements/issues/3570) cpu manager KEP and add the approach for windows cpu manager? then we can come to consensus on how it all works, have somewhere to reference, and then we can move forward here. It doesn't need to follow normal KEP deadlines or be a new stage, just having somewhere to talk about approach and differences between windows and linux will be helpful. WDYT

This sounds like a great approach.  Thanks for taking this to sig-node and coming up with a way forward.  We will work on the requested changes to the KEP to highlight how Windows will work.

**Comment by jsturtevant** [comment]
Followed up with https://github.com/kubernetes/enhancements/pull/4738

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
Comment (kevpar) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1060]: I think `cpu` in the name of both of these is redundant. How about just `mask` and `group`?

Comment (kiashok) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1060]: group seems to be keyword in protobuf already.

**Comment by kiashok** [comment]
> I kinda lean towards one but what do @SergeyKanzhelev and @mrunalp think? It is less code than the[ cgroup driver from cri kep](https://github.com/kubernetes/enhancements/issues/4033) which required one

@haircommander  @SergeyKanzhelev  @mrunalp  @marosset  would you be able to take a second look please? PR has been updated after additional testing with respective changes on containerd/shim and kubelet.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
Comment (marosset) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1056]:
```suggestion
    // Optionally specifies the set of CPUs to affinitize for this container.
```
nit

**Comment by marosset** [comment]
LGTM - had one nit on a comment for a message properly

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
Comment (SergeyKanzhelev) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1062]: still there

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1056]: Can we add some notes on what cri implementation must do when this structure contains incorrect data.

**Comment by SergeyKanzhelev** [comment]
/hold

for https://github.com/kubernetes/enhancements/pull/4738

**Comment by SergeyKanzhelev** [comment]
/hold 

for https://github.com/kubernetes/enhancements/pull/4738 to be merged or alternative KEP started

**Comment by jsturtevant** [comment]
https://github.com/kubernetes/enhancements/pull/4888 has merged so this can go in.  @kiashok could you address the couple comments?

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
Comment (kiashok) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1056]: CRI would basically just forward these values to the shim via oci runtime spec. hcsshim for example will error out while trying to set the affinity values if they are invalid. Are you suggesting we summarize this in the comment here?

cc @jsturtevant

**Review by marosset — APPROVED** [review]
/lgtm

**Pull request merged**
